File: haskell/010-summation_of_primes.hs
-- Summation of primes

-- Problem 10
-- The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

-- Find the sum of all the primes below two million.


main = do
    print (sum $ eratos_sieve [2..1999999])

eratos_sieve :: [Int] -> [Int]
eratos_sieve [] = []
eratos_sieve (x:xs)
    | x * x > last xs = x:xs
    | otherwise = x:eratos_sieve [n | n <- xs, n `mod` x /= 0]

File: julia/010-summation_of_primes.jl
###
# Summation of primes
# Problem 10
#
# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
# Find the sum of all the primes below two million.
###

function eratosthenes_sieve(stop)
    ns = Array(reverse(2:stop))
    primes = []
    while true
        prime = pop!(ns)
        push!(primes, prime)
        if prime > ceil(Integer, √stop)
            return append!(primes, ns)
            break
        end
        filter!(n -> n % prime != 0, ns)
    end
end

result = eratosthenes_sieve(2_000_000 - 1) |> sum

println(result)
